### DISCUSSION TITLE
Updating current screen before `push_screen`?

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#2035] [Q&A] [answered]
Here's a minimal reproduction where a "Login successful!" message should be displayed before moving to the next screen. If you uncomment the `push_screen` line, the message no longer displays. 

I'm guessing this has something to do with concurrency in Textual, but not sure how to approach a possible workaround?

Thanks in advance for any advice.

```python
import time

from textual.app import App, ComposeResult
from textual.containers import Container
from textual.screen import Screen
from textual.widgets import Input, Label

PASSWORD = "textual"


class LoginScreen(Screen):
    def compose(self) -> ComposeResult:
        yield Container(
            Input(placeholder="Password"),
            Label(" "),
        )

    def on_mount(self) -> None:
        self.query_one(Input).focus()

    def on_input_submitted(self, event: Input.Submitted) -> None:
        password: str = event.input.value.strip()
        msg = self.query_one(Label)
        if password == PASSWORD:
            msg.update("Login successful! Loading...")
            # self.app.push_screen("main")
        else:
            msg.update("Incorrect password!")


class MainScreen(Screen):
    def on_mount(self) -> None:
        time.sleep(3)

    def compose(self) -> ComposeResult:
        yield Label("Hello World!")


class ExampleApp(App):
    CSS = (
        "Screen { align: center middle; }"
        "Container { align: center middle; max-width: 100; }"
    )
    SCREENS = {
        "login": LoginScreen,
        "main": MainScreen,
    }

    def on_mount(self) -> None:
        self.push_screen("login")


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Comment by davep** [comment, marked as answer]
Rather than just push the screen right away, I'd do this (assuming `from functools import partial`):

```python
msg.call_after_refresh(partial(self.app.push_screen,"main"))
```

In other words: once the label has fully updated, *then* swap to the main screen.

**Reply by TomJGooding** [reply]
Thanks as always Dave! I wasn't aware of `call_after_refresh`... would you mind explaining what wrapping inside `partial` is doing here?

**Reply by davep** [reply]
So `call_after_refresh` takes a callable to call *after* the refresh happens, but `push_screen` has a parameter so we can't just pass a reference to the method alone, so I use it to make a partial that can be called later along with its parameter.

I could have written it like:

```python
msg.call_after_refresh(lambda: self.app.push_screen("main"))
```

but I generally find Python's `lambda` ugly. ;-)

**Reply by davep** [reply]
@willmcgugan just reminded me that `call_after_refresh` can handle this, so this would work:

```python
msg.call_after_refresh(self.app.push_screen,"main")
```

I forgot that it takes the arguments and makes the `partial` for you.

**Reply by TomJGooding** [reply]
Thank you both for your help!
